[Instructions] cache computed shufflevector properties

Prince781 opened the pull request:
Cache computed properties of a shufflevector mask.

Prince781 commented:
@nikic If I'm not misreading these build logs, the errors look unrelated. Would it be possible for you to assign someone to review this?

nikic reviewed (commented):
I didn't see any compile-time impact when testing this, but probably this needs a fairly specific workload to see benefits. Could you maybe provide some numbers on what you see?

In terms of implementation, I do wonder whether it would make sense to always compute the full properties on construction / mask change, as the lazy initialization makes this somewhat ugly.

Prince781 commented:
Hi @nikic,

We have some internal kernels that make heavy use of very large vectors (256, 512, 1024 elements). We find a 20-40% compile time reduction on these by caching shufflevector mask properties. I think this impact is hard to see with smaller sizes.

Prince781 commented:
The lazy initialization is helpful for our compile time cases, so I'd prefer to keep it. I can refine this patch to be easier to read.

RKSimon commented:
We could consider creating a TTI::ShuffleKindSet mask type and create the entire set in one call in the ShuffleVectorInst constructor - the ShuffleVectorInst::is*Mask calls are all very similar, it shouldn't take much to merge them all into a single analysis loop.

Prince781 commented:
@RKSimon interesting idea. I'll try that out.

Prince781 commented:
@RKSimon why would it be in TTI though? Aren't these target-independent properties?

Prince781 commented:
Nevermind, I see that ShuffleVectorInst methods are used in `BasicTTIImpl.h`

Prince781 committed:
[Instructions] cache computed shufflevector properties

- Cache computed properties of a shufflevector mask.
- Merge several shuffle mask property analyses into a single analysis
  and introduce ShuffleMaskAttrs.
- Compute the properties on shufflevector construction.

Prince781 commented:
Hi all,

Updated the PR to merge shufflevector attribute analyses into a single pass over the mask. I've left other users (like BasicTTIImpl) alone for now, as I don't have a real-world example to motivate changing them.

Prince781 commented:
Improvements on our (@NVIDIA's) end:

before:
```
===-------------------------------------------------------------------------===
                          Pass execution timing report
===-------------------------------------------------------------------------===
  Total Execution Time: 76.5499 seconds (76.5545 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
  72.8495 ( 96.6%)   1.0520 ( 94.9%)  73.9014 ( 96.5%)  73.9059 ( 96.5%)  InstCombinePass #2
   2.1374 (  2.8%)   0.0359 (  3.2%)   2.1733 (  2.8%)   2.1734 (  2.8%)  InstCombinePass #1
   0.0968 (  0.1%)   0.0000 (  0.0%)   0.0968 (  0.1%)   0.0968 (  0.1%)  GVNPass #2
   0.0545 (  0.1%)   0.0120 (  1.1%)   0.0665 (  0.1%)   0.0666 (  0.1%)  SROAPass #1
...
```

after:
```
===-------------------------------------------------------------------------===
                          Pass execution timing report
===-------------------------------------------------------------------------===
  Total Execution Time: 76.5499 seconds (76.5545 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
  72.8495 ( 96.6%)   1.0520 ( 94.9%)  73.9014 ( 96.5%)  73.9059 ( 96.5%)  InstCombinePass #2
   2.1374 (  2.8%)   0.0359 (  3.2%)   2.1733 (  2.8%)   2.1734 (  2.8%)  InstCombinePass #1
   0.0968 (  0.1%)   0.0000 (  0.0%)   0.0968 (  0.1%)   0.0968 (  0.1%)  GVNPass #2
   0.0545 (  0.1%)   0.0120 (  1.1%)   0.0665 (  0.1%)   0.0666 (  0.1%)  SROAPass #1
...
```

after (with [`-instcombine-simplify-vector-elts-depth=1`](https://github.com/llvm/llvm-project/pull/113717)):
```
===-------------------------------------------------------------------------===
                          Pass execution timing report
===-------------------------------------------------------------------------===
  Total Execution Time: 76.5499 seconds (76.5545 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
  72.8495 ( 96.6%)   1.0520 ( 94.9%)  73.9014 ( 96.5%)  73.9059 ( 96.5%)  InstCombinePass #2
   2.1374 (  2.8%)   0.0359 (  3.2%)   2.1733 (  2.8%)   2.1734 (  2.8%)  InstCombinePass #1
   0.0968 (  0.1%)   0.0000 (  0.0%)   0.0968 (  0.1%)   0.0968 (  0.1%)  GVNPass #2
   0.0545 (  0.1%)   0.0120 (  1.1%)   0.0665 (  0.1%)   0.0666 (  0.1%)  SROAPass #1
...
```

Prince781 commented:
Ping.

nikic reviewed (commented):
It may be cleaner to separate out the scalable case into a separate code path, because there is only a single supported scalable shufflevector operation (which is the zero index splat).

RKSimon commented on llvm/lib/IR/Instructions.cpp:
```diff
@@ -1844,6 +1852,106 @@ Constant *ShuffleVectorInst::convertShuffleMaskForBitcode(ArrayRef<int> Mask,
+  ShuffleMaskAttrs MaskAttrs = {};
+
+  // Single-source if uses either LHS or RHS but not both.
+  MaskAttrs.SingleSource = (UsesLHS ^ UsesRHS) && PreservesLength;
```
How much of penalty do we see if we just use:
```
MaskAttrs.SingleSource = !changesLength() && isSingleSourceMask(Mask, Mask.size())
```
etc. for all the other kinds - I don't want to find myself pulling my hair out finding subtle diffs between all of this and the individual static matchers.

Prince781 replied:
I think I could just go ahead and replace the static matchers covered by this implementation, then?

RKSimon replied:
I'd prefer you keep the static matchers - make analyzeMask public and just have it call the individual matchers. Your currently implementation isn't as easy to grok tbh.

Prince781 replied:
Let me give it a try and see if it looks better. If it doesn't, I'll try your suggestion.

Prince781 replied:
Sorry, I posted that comment without reading your reply. I'll try it with the static matchers then.

Prince781 committed:
Revert "[Instructions] cache computed shufflevector properties"

This reverts commit 24a39050595b7f32e7374f67d281d91bfd3cf303.

Prince781 committed:
[Instructions] reimplement shufflevector property caching

Simplify implementation

Prince781 committed:
remove unused fields in ShuffleMaskAttrs

Prince781 commented:
Ping reviewers.

Prince781 committed:
remove mutable

nikic commented on llvm/include/llvm/IR/Instructions.h:
```diff
@@ -1903,6 +1917,27 @@ class ShuffleVectorInst : public Instruction {
+
+  // (Re)compute the shuffle mask attributes.
+  void computeShuffleAttrs() {
+    ShuffleAttrs = {}; // reset
```
Not needed, as you're reassigning all members anyway.

nikic reviewed (approved):
LGTM

Prince781 commented:
Just in case, please hold off on merging as I need to do some more compile time testing.

RKSimon reviewed (approved):
LGTM cheers - @Prince781 hows the compile time checking going?

Prince781 commented:
Hi @RKSimon 

While this improves the original case I was interested in, I'm finding with a newer kernel we recently added that the 8b size increase for ShuffleVectorInst is causing (CPU) caching problems. So I'd like to still hold off on merging.
